seis/sl/slink.go

```diff
@@ -0,0 +1,303 @@
+}
+
+// AddState appends the list of station state information.
+func (s *SLink) AddState(stations ...Station) {
+	s.State = append(s.State, stations...)
+}
+
+// NewSlink returns a SLink pointer for the given server, optional settings can be passed
+
+// CollectFunc is a function run on each returned seedlink packet. It should return a true value
+// to stop collecting data without an error message. A non-nil returned error will also stop
+// collection but with an assumed errored state.
+type CollectFunc func(string, []byte) (bool, error)
+
+// CollectWithContext makes a connection to the seedlink server, recovers initial client information and
+// the sets the connection into streaming mode. Recovered packets are passed to a given function
+// to process, if this function returns a true value or a non-nil error value the collection will
+// stop and the function will return.
+// If a call returns with a timeout error a check is made whether a keepalive is needed or whether
+// the function should return as no data has been received for an extended period of time. It is
+// assumed the calling function will attempt a reconnection with an updated set of options, specifically
+// any start or end time parameters. The Context parameter can be used to to cancel the data collection
+// independent of the function as this may never be called if no appropriate has been received.
+func (s *SLink) CollectWithContext(ctx context.Context, fn CollectFunc) error {
+	ctx, cancel := context.WithCancel(ctx)
+	defer cancel()
+
+	var state State
+	for _, v := range s.State {
+		state.Add(v)
+	}
+
+	list, err := decodeStreams(s.Streams, s.Selectors)
+	if err != nil {
+		return err
+	}
+
+	conn, err := NewConn(s.Server, s.Timeout)
+	if err != nil {
+		return err
+	}
+	defer conn.Close()
+
+	for _, l := range list {
+		if err := conn.CommandStation(l.station, l.network); err != nil {
+			return err
+		}
+
+		if err := conn.CommandSelect(l.selection); err != nil {
+			return err
+		}
+
+		sequence, starttime := s.Sequence, s.Start
+		if v := state.Find(Station{Network: l.network, Station: l.station}); v != nil {
+			sequence, starttime = v.Sequence, v.Timestamp
+		}
+
+		switch {
+		case !s.End.IsZero():
+			if err := conn.CommandTime(s.Start, s.End); err != nil {
+				return err
+			}
+			// there may be a sequence number
+		case !(sequence < 0):
+			//convert the next sequence number into uppercase hex
+			seq := fmt.Sprintf("%06X", (s.Sequence+1)&0xffffff)
+			if err := conn.CommandData(seq, starttime); err != nil {
+				return err
+			}
+		default:
+			// or check a possible start time
+			if err := conn.CommandTime(starttime, time.Time{}); err != nil {
+				return err
+			}
+		}
+	}
+	if err := conn.CommandEnd(); err != nil {
+		return err
+	}
+
+	last := time.Now()
+
+loop:
+	for {
+		select {
+		case <-ctx.Done():
+			break loop
+		default:
+			switch pkt, err := conn.Collect(); {
+			case err != nil:
+				switch err := err.(type) {
```

Comment:
would not assign type to the err variable itself

Reply:
this gets knarly, basically it will cast the return err into the type value used in the switch. rather than the actual err type.